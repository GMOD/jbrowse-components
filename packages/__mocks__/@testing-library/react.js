import { StylesProvider } from '@material-ui/core/styles'
import React from 'react'

const react = jest.requireActual('@testing-library/react')
const generateClassName = (rule, styleSheet) =>
  `${styleSheet.options.classNamePrefix}-${rule.key}`
const render = args => {
  return react.render(
    <StylesProvider generateClassName={generateClassName}>
      {args}
    </StylesProvider>,
  )
}

module.exports = { ...react, render }
