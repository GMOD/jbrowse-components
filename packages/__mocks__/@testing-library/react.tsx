import React, { Suspense } from 'react'
import { createTheme, ThemeProvider } from '@mui/material/styles'

const react = jest.requireActual('@testing-library/react')
const render = (args: React.ReactNode) => {
  return react.render(
    <Suspense fallback={null}>
      <ThemeProvider theme={createTheme()}>{args}</ThemeProvider>
    </Suspense>,
  )
}

module.exports = { ...react, render }
