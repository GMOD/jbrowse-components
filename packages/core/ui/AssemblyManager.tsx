import React from 'react'
import { makeStyles } from '@material-ui/core/styles'
import Button from '@material-ui/core/Button'
import Dialog from '@material-ui/core/Dialog'
import DialogActions from '@material-ui/core/DialogActions'
import DialogContent from '@material-ui/core/DialogContent'
import DialogTitle from '@material-ui/core/DialogTitle'
import Table from '@material-ui/core/Table'
import TableBody from '@material-ui/core/TableBody'
import TableCell from '@material-ui/core/TableCell'
import TableContainer from '@material-ui/core/TableContainer'
import TableHead from '@material-ui/core/TableHead'
import TableRow from '@material-ui/core/TableRow'
import Paper from '@material-ui/core/Paper'

const useStyles = makeStyles(() => ({
  titleBox: {
    color: '#FFFFFF',
    backgroundColor: '#0D233F',
  },
  table: {
    minWidth: 650,
  },
}))

// remember to make observable
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function AssemblyTable(props: any) {
  const { rootModel } = props
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const rows = rootModel.jbrowse.assemblies.map((assembly: any) => {
    const { name } = assembly
    return (
      <TableRow>
        <TableCell>{name}</TableCell>
      </TableRow>
    )
  })

  return (
    <TableContainer component={Paper}>
      <TableHead>
        <TableRow>
          <TableCell>Name</TableCell>
          <TableCell>Aliases</TableCell>
        </TableRow>
      </TableHead>
      <Table>
        <TableBody>{rows}</TableBody>
      </Table>
    </TableContainer>
  )
}

const AssemblyManager = ({
  rootModel,
  open,
  onClose,
}: {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  rootModel: any
  open: boolean
  onClose: Function
}) => {
  const classes = useStyles()

  return (
    <Dialog open={open}>
      <DialogTitle className={classes.titleBox}>Assembly Manager</DialogTitle>
      <DialogContent>
        <AssemblyTable rootModel={rootModel} />
      </DialogContent>
      <DialogActions>
        <Button
          color="secondary"
          variant="contained"
          onClick={() => {
            onClose(false)
          }}
        >
          OK
        </Button>
        <Button
          color="secondary"
          variant="contained"
          onClick={() => {
            onClose(false)
          }}
        >
          Cancel
        </Button>
      </DialogActions>
    </Dialog>
  )
}

// remember to make observable
export default AssemblyManager
