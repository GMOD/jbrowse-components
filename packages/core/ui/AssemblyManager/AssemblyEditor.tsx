import React from 'react'
import { observer } from 'mobx-react'
import { ConfigurationEditor } from '../configEditor'

const AssemblyEditor = observer(
  ({
    assembly,
  }: {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    assembly: any
  }) => {
    return <ConfigurationEditor target={assembly} />
  },
)

export default AssemblyEditor
