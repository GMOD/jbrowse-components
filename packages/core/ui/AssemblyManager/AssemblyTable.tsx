import React from 'react'
import { observer } from 'mobx-react'
import { makeStyles } from '@material-ui/core/styles'
import Button from '@material-ui/core/Button'
import Table from '@material-ui/core/Table'
import TableBody from '@material-ui/core/TableBody'
import TableCell from '@material-ui/core/TableCell'
import TableContainer from '@material-ui/core/TableContainer'
import TableHead from '@material-ui/core/TableHead'
import TableRow from '@material-ui/core/TableRow'
import Paper from '@material-ui/core/Paper'
import CreateIcon from '@material-ui/icons/Create'
import DeleteIcon from '@material-ui/icons/Delete'

// local
import { readConfObject } from '../../configuration'

const useStyles = makeStyles(() => ({
  table: {
    minWidth: 650,
  },
  buttonCell: {
    padding: 3,
  },
  button: {
    display: 'inline-block',
    padding: 0,
    minHeight: 0,
    minWidth: 0,
  },

  dialogContent: {
    width: 600,
  },
}))

const AssemblyTable = observer(
  ({
    rootModel,
    setIsAssemblyBeingEdited,
    setAssemblyBeingEdited,
  }: {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    rootModel: any
    setIsAssemblyBeingEdited(arg: boolean): void
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    setAssemblyBeingEdited(arg: any): void
  }) => {
    const classes = useStyles()

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const rows = rootModel.jbrowse.assemblies.map((assembly: any) => {
      const name = readConfObject(assembly, 'name')
      const aliases = readConfObject(assembly, 'aliases')
      return (
        <TableRow key={name}>
          <TableCell>{name}</TableCell>
          <TableCell>{aliases.toString()}</TableCell>
          <TableCell className={classes.buttonCell}>
            <Button
              className={classes.button}
              onClick={() => {
                setIsAssemblyBeingEdited(true)
                setAssemblyBeingEdited(assembly)
              }}
            >
              <CreateIcon color="primary" />
            </Button>
          </TableCell>
          <TableCell className={classes.buttonCell}>
            <Button className={classes.button}>
              <DeleteIcon color="error" />
            </Button>
          </TableCell>
        </TableRow>
      )
    })

    return (
      <TableContainer component={Paper}>
        <Table>
          <TableHead>
            <TableRow>
              <TableCell>Name</TableCell>
              <TableCell>Aliases</TableCell>
              <TableCell>Actions</TableCell>
            </TableRow>
          </TableHead>
          <TableBody>{rows}</TableBody>
        </Table>
      </TableContainer>
    )
  },
)

export default AssemblyTable
