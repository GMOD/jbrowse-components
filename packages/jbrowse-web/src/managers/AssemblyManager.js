import { readConfObject } from '@gmod/jbrowse-core/configuration'

export default class AssemblyManager {
  constructor(rpcManager, rootModel, opts = {}) {
    rpcManager.assemblyManager = this
    this.rpcManager = rpcManager
    this.refNameMaps = new Map()
    this.updateAssemblyData(rootModel, opts)
  }

  updateAssemblyData(rootModel, opts = {}) {
    const rootConfig = rootModel.configuration
    this.assemblyData = new Map()
    for (const assemblyConfig of rootConfig.assemblies) {
      const assemblyName = readConfObject(assemblyConfig, 'assemblyName')
      const assemblyInfo = {}
      if (assemblyConfig.sequence)
        assemblyInfo.sequence = assemblyConfig.sequence
      const refNameAliasesConf = readConfObject(
        assemblyConfig,
        'refNameAliases',
      )
      if (refNameAliasesConf) assemblyInfo.refNameAliases = refNameAliasesConf
      const aliases = readConfObject(assemblyConfig, 'aliases')
      assemblyInfo.aliases = aliases
      this.assemblyData.set(assemblyName, assemblyInfo)
      aliases.forEach((assemblyAlias, idx) => {
        const newAliases = [
          ...aliases.slice(0, idx),
          ...aliases.slice(idx + 1),
          assemblyName,
        ]
        this.assemblyData.set(assemblyAlias, {
          ...assemblyInfo,
          aliases: newAliases,
        })
      })
    }
    rootModel.connections.forEach(connection => {
      connection.assemblies.forEach(assembly => {
        const { assemblyName } = assembly
        if (!this.assemblyData.has(assemblyName)) {
          const assemblyInfo = {}
          assemblyInfo.sequence = assembly.sequence
          assemblyInfo.refNameAliases = assembly.refNameAliases
          this.assemblyData.set(assemblyName, assemblyInfo)
        } else {
          if (
            !this.assemblyData.get(assemblyName).refNameAliases &&
            assembly.refNameAliases
          ) {
            this.assemblyData.get(assemblyName).refNameAliases = readConfObject(
              assembly.refNameAliases,
            )
            this.assemblyData.get(assemblyName).aliases.forEach(alias => {
              this.assemblyData.get(alias).refNameAliases = readConfObject(
                assembly.refNameAliases,
              )
            })
          }
          if (
            (!this.assemblyData.get(assemblyName).sequence &&
              assembly.sequence) ||
            assembly.defaultSequence
          ) {
            this.assemblyData.get(assemblyName).sequence = assembly.sequence

            this.assemblyData.get(assemblyName).aliases.forEach(alias => {
              this.assemblyData.get(alias).sequence = assembly.sequence
            })
          }
        }
      })
    })

    this.setDisplayedRegions(rootModel, opts)
  }

  async setDisplayedRegions(rootModel, opts = {}) {
    for (const view of rootModel.views) {
      const assemblyName = view.displayRegionsFromAssemblyName
      if (assemblyName && this.assemblyData.get(assemblyName).sequence) {
        // eslint-disable-next-line no-await-in-loop
        const displayedRegions = await this.getRegionsForAssembly(
          assemblyName,
          opts,
        )
        view.setDisplayedRegions(displayedRegions, true)
      }
    }
  }

  /**
   * Return an object like { refName1: ['alias1', 'alias2'], refName2:
   * ['alias3', 'alias4'] } that describes the reference names and aliases of
   * the given assembly. Returns empty object if the assembly is not found.
   * @param {string} assemblyName The name of an assembly
   * @returns {Promise<object>} Object like { refName1: ['alias1', 'alias2'],
   * refName2: ['alias3', 'alias4'] }
   */
  async getRefNameAliases(assemblyName, opts = {}) {
    const refNameAliases = {}
    const assemblyConfig = this.assemblyData.get(assemblyName)
    if (assemblyConfig.refNameAliases) {
      // eslint-disable-next-line no-await-in-loop
      const adapterRefNameAliases = await this.rpcManager.call(
        assemblyConfig.refNameAliases.adapter.configId,
        'getRefNameAliases',
        {
          sessionId: assemblyName,
          adapterType: assemblyConfig.refNameAliases.adapter.type,
          adapterConfig: assemblyConfig.refNameAliases.adapter,
          signal: opts.signal,
        },
        { timeout: 1000000 },
      )
      adapterRefNameAliases.forEach(alias => {
        refNameAliases[alias.refName] = alias.aliases
      })
    }
    return refNameAliases
  }

  /**
   * Add a map with entries like (refName alias -> refName used by this adapter)
   * to the `refNameMaps` attribute
   * @param {adapterConf} adapterConf Configuration model of an adapter
   * @param {string} assemblyName Assembly to use for aliasing
   */
  async addRefNameMapForAdapter(adapterConf, assemblyName, opts = {}) {
    const refNameMap = new Map()

    const refNameAliases = await this.getRefNameAliases(assemblyName, opts)

    const refNames = await this.rpcManager.call(
      readConfObject(adapterConf, 'configId'),
      'getRefNames',
      {
        sessionId: assemblyName,
        adapterType: readConfObject(adapterConf, 'type'),
        adapterConfig: adapterConf,
        signal: opts.signal,
      },
      { timeout: 1000000 },
    )
    refNames.forEach(refName => {
      if (refNameAliases[refName])
        refNameAliases[refName].forEach(refNameAlias => {
          refNameMap.set(refNameAlias, refName)
        })
      else
        Object.keys(refNameAliases).forEach(configRefName => {
          if (refNameAliases[configRefName].includes(refName)) {
            refNameMap.set(configRefName, refName)
            refNameAliases[configRefName].forEach(refNameAlias => {
              if (refNameAlias !== refName)
                refNameMap.set(refNameAlias, refName)
            })
          }
        })
    })

    this.refNameMaps.set(readConfObject(adapterConf, 'configId'), refNameMap)
  }

  /**
   * Return cached refNameMap for the adapter, creating and storing it first if
   * needed
   * @param {adapterConf} adapterConf Configuration model of an adapter
   * @param {string} assemblyName Assembly to use for aliasing
   * @returns {Map} See `addRefNameMapForAdapter` for example Map, or an empty
   * map if the adapter is not found.
   */
  async getRefNameMapForAdapter(adapterConf, assemblyName, opts = {}) {
    const configId = readConfObject(adapterConf, 'configId')
    if (!this.refNameMaps.has(configId))
      await this.addRefNameMapForAdapter(adapterConf, assemblyName, opts)
    return this.refNameMaps.get(configId)
  }

  /**
   * Clear all stored refNameMaps
   */
  clear() {
    this.refNameMaps = new Map()
  }

  async getRegionsForAssembly(assemblyName, opts = {}) {
    const assembly = this.assemblyData.get(assemblyName)
    if (assembly) {
      const adapterConfig = readConfObject(assembly.sequence, 'adapter')
      // eslint-disable-next-line no-await-in-loop
      const adapterRegions = await this.rpcManager.call(
        assembly.configId,
        'getRegions',
        {
          sessionId: assemblyName,
          adapterType: adapterConfig.type,
          adapterConfig,
          signal: opts.signal,
        },
        { timeout: 1000000 },
      )
      const adapterRegionsWithAssembly = adapterRegions.map(adapterRegion => ({
        ...adapterRegion,
        assemblyName,
      }))
      return adapterRegionsWithAssembly
    }
    return undefined
  }
}
