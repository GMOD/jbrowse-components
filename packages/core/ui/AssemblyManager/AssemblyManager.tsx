import React, { useState } from 'react'
import { observer } from 'mobx-react'
import { makeStyles } from '@material-ui/core/styles'
import Button from '@material-ui/core/Button'
import Dialog from '@material-ui/core/Dialog'
import DialogActions from '@material-ui/core/DialogActions'
import DialogContent from '@material-ui/core/DialogContent'
import DialogTitle from '@material-ui/core/DialogTitle'
import AddIcon from '@material-ui/icons/Add'

// Add Form imports
import TextField from '@material-ui/core/TextField'
import { ConfigurationEditor } from '../configEditor'

import AssemblyTable from './AssemblyTable'

const useStyles = makeStyles(() => ({
  titleBox: {
    color: '#FFFFFF',
    backgroundColor: '#0D233F',
  },
  dialogContent: {
    width: 600,
  },
}))

const AssemblyAddForm = observer(
  ({ setFormOpen }: { setFormOpen: Function }) => {
    return (
      <>
        <form noValidate autoComplete="off">
          <TextField
            id="assembly-name"
            label="Assembly Name"
            variant="outlined"
          />
          <TextField id="aliases" label="Aliases" variant="outlined" />
        </form>
        <Button
          variant="contained"
          color="secondary"
          onClick={() => setFormOpen(false)}
        >
          Close
        </Button>
      </>
    )
  },
)

const AssemblyEditor = observer(
  ({
    assembly,
  }: {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    assembly: any
  }) => {
    return (
      <>
        <TextField>Rename field</TextField>
        <ConfigurationEditor target={assembly} />
      </>
    )
  },
)

const AssemblyManager = observer(
  ({
    rootModel,
    open,
    onClose,
  }: {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    rootModel: any
    open: boolean
    onClose: Function
  }) => {
    const classes = useStyles()
    const [isFormOpen, setFormOpen] = useState(false)
    const [assemblyBeingEdited, setAssemblyBeingEdited] = useState()

    const showAssemblyTable = !isFormOpen && !assemblyBeingEdited

    return (
      <Dialog open={open}>
        <DialogTitle className={classes.titleBox}>Assembly Manager</DialogTitle>
        <DialogContent>
          <div className={classes.dialogContent}>
            {showAssemblyTable ? (
              <>
                <Button
                  variant="contained"
                  color="secondary"
                  startIcon={<AddIcon />}
                  onClick={() => {
                    setFormOpen(true)
                  }}
                >
                  Add New Assembly
                </Button>
                <AssemblyTable
                  rootModel={rootModel}
                  setAssemblyBeingEdited={setAssemblyBeingEdited}
                />
              </>
            ) : null}
            {assemblyBeingEdited ? (
              <AssemblyEditor assembly={assemblyBeingEdited} />
            ) : null}
            {isFormOpen ? <AssemblyAddForm setFormOpen={setFormOpen} /> : null}
          </div>
        </DialogContent>
        <DialogActions>
          {showAssemblyTable ? (
            <Button
              color="secondary"
              variant="contained"
              onClick={() => {
                onClose(false)
              }}
            >
              Return
            </Button>
          ) : null}
        </DialogActions>
      </Dialog>
    )
  },
)

export default AssemblyManager
