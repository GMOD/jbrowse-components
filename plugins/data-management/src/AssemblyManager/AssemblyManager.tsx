import React, { useState } from 'react'
import { observer } from 'mobx-react'
import { makeStyles } from '@material-ui/core/styles'
import Button from '@material-ui/core/Button'
import Dialog from '@material-ui/core/Dialog'
import DialogActions from '@material-ui/core/DialogActions'
import DialogContent from '@material-ui/core/DialogContent'
import DialogTitle from '@material-ui/core/DialogTitle'
import AddIcon from '@material-ui/icons/Add'
import ArrowBackIosIcon from '@material-ui/icons/ArrowBackIos'
import { IconButton } from '@material-ui/core/'

import AssemblyTable from './AssemblyTable'
import AssemblyAddForm from './AssemblyAddForm'
import AssemblyEditor from './AssemblyEditor'

const useStyles = makeStyles(theme => ({
  titleBox: {
    color: '#fff',
    backgroundColor: theme.palette.primary.main,
    textAlign: 'center',
  },
  dialogContent: {
    width: 600,
  },
  backButton: {
    color: '#fff',
    position: 'absolute',
    left: theme.spacing(4),
    top: theme.spacing(4),
  },
}))

const AssemblyManager = observer(
  ({
    rootModel,
    open,
    onClose,
  }: {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    rootModel: any
    open: boolean
    onClose: Function
  }) => {
    const classes = useStyles()
    const [isFormOpen, setFormOpen] = useState(false)
    const [isAssemblyBeingEdited, setIsAssemblyBeingEdited] = useState(false)
    const [assemblyBeingEdited, setAssemblyBeingEdited] = useState()

    const showAssemblyTable = !isFormOpen && !isAssemblyBeingEdited

    return (
      <Dialog open={open}>
        <DialogTitle className={classes.titleBox}>
          {showAssemblyTable ? 'Assembly manager' : null}
          {isFormOpen ? (
            <>
              <IconButton
                aria-label="back"
                className={classes.backButton}
                onClick={() => setFormOpen(false)}
              >
                <ArrowBackIosIcon />
              </IconButton>
              Add new assembly
            </>
          ) : null}
          {isAssemblyBeingEdited ? (
            <>
              <IconButton
                aria-label="back"
                className={classes.backButton}
                onClick={() => setIsAssemblyBeingEdited(false)}
              >
                <ArrowBackIosIcon />
              </IconButton>
              {returnAssemblyName(assemblyBeingEdited)}
            </>
          ) : null}
        </DialogTitle>
        <DialogContent>
          <div className={classes.dialogContent}>
            {showAssemblyTable ? (
              <AssemblyTable
                rootModel={rootModel}
                setIsAssemblyBeingEdited={setIsAssemblyBeingEdited}
                setAssemblyBeingEdited={setAssemblyBeingEdited}
              />
            ) : null}
            {isAssemblyBeingEdited ? (
              <AssemblyEditor assembly={assemblyBeingEdited} />
            ) : null}
            {isFormOpen ? (
              <AssemblyAddForm
                rootModel={rootModel}
                setFormOpen={setFormOpen}
              />
            ) : null}
          </div>
        </DialogContent>
        <DialogActions>
          {showAssemblyTable ? (
            <>
              <Button
                color="secondary"
                variant="contained"
                onClick={() => {
                  onClose(false)
                }}
              >
                Return
              </Button>
              <Button
                variant="contained"
                color="secondary"
                startIcon={<AddIcon />}
                onClick={() => {
                  setFormOpen(true)
                }}
              >
                Add New Assembly
              </Button>
            </>
          ) : null}
        </DialogActions>
      </Dialog>
    )
  },
)

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function returnAssemblyName(assembly: any) {
  if (assembly !== undefined) {
    return assembly.name
  }
  return null
}

export default AssemblyManager
