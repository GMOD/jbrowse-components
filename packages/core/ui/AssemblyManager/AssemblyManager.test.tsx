import React from 'react'
import { render } from '@testing-library/react'

import AssemblyManager from './AssemblyManager'

const mockRootModel = {
  jbrowse: {
    assemblies: [
      {
        name: 'testAssembly',
        sequence: {
          type: 'testSequenceTrack',
          trackId: '',
          adapter: {
            type: 'testSeqAdapter',
            twoBitLocation: {
              uri: 'test.2bit',
            },
          },
        },
      },
    ],
  },
}

describe('AsssemblyManager GUI', () => {
  it('renders succesfully', () => {
    const { getByText } = render(
      <AssemblyManager rootModel={mockRootModel} open onClose={() => {}} />,
    )
    expect(getByText('Assembly Manager')).toBeTruthy()
  })
})
