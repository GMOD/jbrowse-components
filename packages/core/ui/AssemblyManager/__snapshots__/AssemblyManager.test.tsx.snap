// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AsssemblyManager GUI renders 1`] = `
<div
  aria-hidden="true"
/>
`;
